eth/beacon/state_machines/base.py

```diff
@@ -193,9 +222,562 @@ def active_state(self) -> ActiveState:
+                cycle_length,
+            )
+
+        return active_state.copy(
+            recent_block_hashes=recent_block_hashes,
+            pending_attestations=(
+                active_state.pending_attestations + block.attestations
+            ),
+        )
+
+    @classmethod
+    def compute_cycle_transitions(
+            cls,
+            crystallized_state: CrystallizedState,
+            active_state: ActiveState) -> Tuple[CrystallizedState, ActiveState]:
+        # TODO: it's a stub
+        return crystallized_state, active_state
+
+    #
+    #
+    # Proposer APIs
+    #
+    #
+    def propose_block(
+        self,
+        crystallized_state: CrystallizedState,
+        active_state: ActiveState,
+        block: BaseBeaconBlock,
+        shard_id: int,
+        shard_block_hash: Hash32,
+        chaindb: BaseBeaconBlock,
+        config: BeaconConfig,
+        private_key: int
+    ) -> Tuple[BaseBeaconBlock, CrystallizedState, ActiveState, 'AttestationRecord']:
+        """
+        Propose the given block.
+        """
+        block, post_crystallized_state, post_active_state = self.process_block(
+            crystallized_state,
+            active_state,
+            block,
+            chaindb,
+            config,
+        )
+
+        # Set state roots
+        block.copy(
+            crystallized_state_root=post_crystallized_state.hash,
+            active_state_root=post_active_state,
+        )
+
+        proposer_attestation = self.attest_proposed_block(
+            post_crystallized_state,
+            post_active_state,
+            block,
+            shard_id,
+            shard_block_hash,
+            chaindb,
+            config.CYCLE_LENGTH,
+            private_key,
+        )
+        return block, post_crystallized_state, post_active_state, proposer_attestation
+
+    def _update_the_states(self,
+                           crystallized_state: CrystallizedState,
+                           active_state: ActiveState) -> None:
+        self._crytallized_state = crystallized_state
+        self._active_state = active_state
+
+    def attest_proposed_block(self,
+                              post_crystallized_state: CrystallizedState,
+                              post_active_state: ActiveState,
+                              block: BaseBeaconBlock,
+                              shard_id: int,
+                              shard_block_hash: Hash32,
+                              chaindb: BaseBeaconBlock,
+                              cycle_length: int,
+                              private_key: int) -> 'AttestationRecord':
+        """
+        Return the initial attestation by the block proposer.
+
+        The proposer broadcasts zir attestation with the proposed block.
+        """
+        block_committees_info = get_block_committees_info(
+            block,
+            post_crystallized_state,
+            cycle_length,
+        )
+        # Vote
+        attester_bitfield = set_voted(
```

Comment:
Not for this PR, but it might be nice to create a class to represent the voting bitfield

```python
class VoteBitfield(bytes):
    @classmethod
    def create_empty(cls, size):
        ...

    def set_voted(self, idx):
        ...

    def get_vote(self, idx):
        ...
```

I don't necesarrily think those are the best APIs, just meant to be illustrative.

API for interacting with it could be similar to the `Receipt.bloom_filter`.  Also, might be nice for the class to be immutable so that each method call returns the updated bitfield, allowing you to chain multiple calls.